storage/src/main/java/zipkin2/storage/kafka/KafkaSpanStore.java

```diff
@@ -13,286 +13,332 @@
-        traceStoreStream.store(TRACES_STORE_NAME, QueryableStoreTypes.keyValueStore());
-    ReadOnlyKeyValueStore<Long, Set<String>> traceIdsByTsStore =
-        traceStoreStream.store(SPAN_IDS_BY_TS_STORE_NAME, QueryableStoreTypes.keyValueStore());
-    return new GetTracesCall(tracesStore, traceIdsByTsStore, request, minTracesStored);
+    return new GetTracesCall(traceStoreStream, request);
   }
 
   @Override
   public Call<List<Span>> getTrace(String traceId) {
-    ReadOnlyKeyValueStore<String, List<Span>> traceStore =
-        traceStoreStream.store(TRACES_STORE_NAME, QueryableStoreTypes.keyValueStore());
-    return new GetTraceCall(traceStore, traceId);
+    return new GetTraceCall(traceStoreStream, traceId);
   }
 
   @Deprecated @Override public Call<List<String>> getServiceNames() {
-    ReadOnlyKeyValueStore<String, String> serviceStore =
-        traceStoreStream.store(SERVICE_NAMES_STORE_NAME, QueryableStoreTypes.keyValueStore());
-    return new GetServiceNamesCall(serviceStore);
+    return new GetServiceNamesCall(traceStoreStream);
   }
 
   @Deprecated @Override public Call<List<String>> getSpanNames(String serviceName) {
-    ReadOnlyKeyValueStore<String, Set<String>> spanNamesStore =
-        traceStoreStream.store(SPAN_NAMES_STORE_NAME, QueryableStoreTypes.keyValueStore());
-    return new GetSpanNamesCall(spanNamesStore, serviceName);
+    return new GetSpanNamesCall(traceStoreStream, serviceName);
   }
 
   @Override public Call<List<String>> getRemoteServiceNames(String serviceName) {
-    ReadOnlyKeyValueStore<String, Set<String>> remoteServiceNamesStore =
-        traceStoreStream.store(REMOTE_SERVICE_NAMES_STORE_NAME,
-            QueryableStoreTypes.keyValueStore());
-    return new GetRemoteServiceNamesCall(remoteServiceNamesStore, serviceName);
+    return new GetRemoteServiceNamesCall(traceStoreStream, serviceName);
   }
 
   @Override public Call<List<DependencyLink>> getDependencies(long endTs, long lookback) {
-    ReadOnlyWindowStore<Long, DependencyLink> dependenciesStore =
-        dependencyStoreStream.store(DEPENDENCIES_STORE_NAME,
-            QueryableStoreTypes.windowStore());
-    return new GetDependenciesCall(endTs, lookback, dependenciesStore);
+    return new GetDependenciesCall(dependencyStoreStream, endTs, lookback);
   }
 
-  static class GetServiceNamesCall extends KafkaStreamsStoreCall<List<String>> {
-    ReadOnlyKeyValueStore<String, String> serviceStore;
+  static class GetServiceNamesCall extends Call.Base<List<String>> {
+    static final Gson GSON = new Gson();
 
-    GetServiceNamesCall(ReadOnlyKeyValueStore<String, String> serviceStore) {
-      this.serviceStore = serviceStore;
+    final KafkaStreams traceStoreStream;
+
+    GetServiceNamesCall(KafkaStreams traceStoreStream) {
+      this.traceStoreStream = traceStoreStream;
+    }
+
+    @Override protected List<String> doExecute() throws IOException {
+      return traceStoreStream.allMetadataForStore(SERVICE_NAMES_STORE_NAME)
```

Comment:
generalize a scatter-gather call so it can be optimized independently. in fact, copy/paste the armeria client call from elasticsearch also (we may at some point extract a separate armeria support lib I think..)

Reply:
> generalize a scatter-gather call

done :) will check elasticsearch call now.